lib/events.js

```diff
@@ -647,3 +649,79 @@ function once(emitter, name) {
     emitter.once(name, eventListener);
   });
 }
+
+const AsyncIteratorPrototype = ObjectGetPrototypeOf(
+  ObjectGetPrototypeOf(async function* () {}).prototype);
+
+function createIterResult(value, done) {
+  return { value, done };
+}
+
+function on(emitter, event) {
+  const unconsumedEvents = [];
+  const unconsumedPromises = [];
+  let error = null;
+
+  const iterator = ObjectSetPrototypeOf({
+    next() {
+      if (error) {
+        return Promise.reject(error);
+      }
+
+      const value = unconsumedEvents.shift();
+      if (value) {
+        return Promise.resolve(createIterResult(value, false));
+      }
+
+      return new Promise(function(resolve, reject) {
+        unconsumedPromises.push({ resolve, reject });
+      });
+    },
+
+    return() {
+      emitter.removeListener(event, eventHandler);
+      emitter.removeListener('error', errorHandler);
+
+      return Promise.resolve(createIterResult(undefined, true));
+    },
+
+    throw(err) {
+      if (!err || !(err instanceof Error)) {
+        throw new ERR_INVALID_ARG_TYPE('EventEmitter.AsyncIterator',
+                                       'Error', err);
+      }
+      error = err;
+      emitter.removeListener(event, eventHandler);
+      emitter.removeListener('error', errorHandler);
+    },
+
+    [Symbol.asyncIterator]() {
+      return this;
+    }
+  }, AsyncIteratorPrototype);
+
+  emitter.on(event, eventHandler);
+  emitter.on('error', errorHandler);
+
+  return iterator;
+
+  function eventHandler(...args) {
+    const promise = unconsumedPromises.shift();
+    if (promise) {
+      promise.resolve(createIterResult(args, false));
+    } else {
+      unconsumedEvents.push(args);
+    }
+  }
+
+  function errorHandler(err) {
+    const promise = unconsumedPromises.shift();
```

Comment:
There is an opportunity here to actually reject all `unconsumedPromises` instead of just the first one.

This will create more deterministic behavior with respect to time.

For example

```
const e = new EventEmitter()
const itr = on(e, 'foo')

const p1 = itr.next() // REJECTED
const p2 = itr.next() // PERMANENTLY PENDING
e.emit('error', new Error())
const p3 = itr.next() // REJECTED
```

```
const e = new EventEmitter()
const itr = on(e, 'foo')

const p1 = itr.next() // REJECTED
e.emit('error', new Error())
const p2 = itr.next() // REJECTED
const p3 = itr.next() // REJECTED
```

Reply:
Implemented.